### PR TITLE
Add some public properties and methods to `Operator2`

### PR DESCRIPTION
**Pull request by mudit2812**
Added the following:
* [x] `arithmetic_depth`
* [x] `is_verified_hermitian`
* [x] `label`
* [x] `pow`
* [x] `has_adjoint`
* [x] `adjoint`
* [x] `map_wires`
* [x] `simplify`

Skeleton tests were added using Cursor. I reviewed, expanded, and refactored as necessary before pushing

[sc-120430]

### CODE REVIEW

File: pennylane/operation2.py [hunk 1]
```diff
@@ -495,17 +500,247 @@ def ndim_params(self) -> tuple[int]:
...
+            .Operator2: new operator
+        """
+        # pylint: disable=protected-access
+        new_op = copy(self)
+
+        for n, wires in self.wire_args.items():
+            leaves, tree = flatten(wires, is_leaf=lambda w: isinstance(w, Wires))
+            mapped_leaves = [Wires([wire_map.get(w, w) for w in leaf]) for leaf in leaves]
+            new_wires = unflatten(mapped_leaves, tree)
+            new_op._bound_args.arguments[n] = new_wires
```
Comment (albi3ro) [inline pennylane/operation2.py:498]: Do we need to do `copy(self)`, or could we just update the arguments then call again?

Also, do we also need to map hybrid args that have operators in them?

File: tests/operation2/test_operation2.py [hunk 2]
```diff
@@ -890,3 +886,365 @@ def test_deepcopy(self):
...
+
+            def __init__(self, wires, ctrl_wires):
+                super().__init__(Wires(wires), Wires(ctrl_wires))
+
+        op = TwoWireOp(wires=[0, 1], ctrl_wires=[2])
+        new_op = op.map_wires({0: "a", 1: "b", 2: "c"})
+
+        assert new_op == TwoWireOp(wires=["a", "b"], ctrl_wires=["c"])
+
+    def test_map_wires_pytree_hybrid_wires(self):
```
Comment (albi3ro) [inline tests/operation2/test_operation2.py:1160]: Test map wires with hybrid operator input?

File: pennylane/operation2.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/operation2.py:498]: Thanks for catching the hybrid args thing. I chose to `copy(self)` because it lets us bypass the constructor.

File: tests/operation2/test_operation2.py (hunk 2 above)
Comment (mudit2812) [reply tests/operation2/test_operation2.py:1160]: [469855b](https://github.com/PennyLaneAI/pennylane/pull/9526/commits/469855bd42131294a55ca0b5fa75751627868cd0)

File: pennylane/operation2.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/operation2.py:498]: [469855b](https://github.com/PennyLaneAI/pennylane/pull/9526/commits/469855bd42131294a55ca0b5fa75751627868cd0)

Comment (albi3ro) [reply pennylane/operation2.py:498]: Why do we want to bypass the constructor? Wouldn't hitting the constructor be a good thing?

Comment (mudit2812) [reply pennylane/operation2.py:498]: All the inputs are already canonicalized so no point in re-initializing everything

File: pennylane/ops/functions/equal.py [hunk 3]
```diff
@@ -384,37 +386,17 @@ def _equal_operator2(
...
-
-        if check_trainability:
-            grad1 = "trainable" if math.requires_grad(dval1) else "not trainable"
-            grad2 = "trainable" if math.requires_grad(dval2) else "not trainable"
-            if grad1 != grad2:
-                return (
-                    f"op1 and op2 differ in trainability for '{dname}'.\n"
-                    f"op1.{dname} is {grad1} and op2.{dname} is {grad2}."
-                )
+        res = _check_dynamic_value(
```
Comment (astralcai) [inline pennylane/ops/functions/equal.py:419]: This seems to be changes that belonged to the upstream PR. Can we merge from it again?

Comment (mudit2812) [reply pennylane/ops/functions/equal.py:419]: odd 🤔

File: pennylane/operation2.py [hunk 4]
```diff
@@ -414,190 +846,10 @@ def _check_batching(self):
 
             self._batch_size = first_dims[0]
 
-    # ------------------ Public properties ------------------
```
Comment (mudit2812) [inline pennylane/operation2.py:418]: All the removed lines in this chunk were just moved above the private utils

File: pennylane/operation2.py [hunk 5]
```diff
@@ -165,6 +168,435 @@ def __init__(self, *args, **kwargs):
...
+
+    @property
+    def hybrid_args(self) -> dict[str, Any]:
+        """Dictionary mapping hybrid argument names to their values."""
+        return {name: self.arguments[name] for name in self.hybrid_argnames}
+
+    @property
+    def compilable_args(self) -> dict[str, Any]:
+        """Dictionary mapping compilable argument names to their values."""
+        return {name: self.arguments[name] for name in self.compilable_argnames}
```
Comment (astralcai) [inline pennylane/operation2.py:204]: I'm wondering if we actually need them to be public. They all look quite internal.

File: pennylane/operation2.py [hunk 6]
```diff
@@ -165,6 +168,435 @@ def __init__(self, *args, **kwargs):
...
+            self._check_batching()
+        return self._ndim_params
+
+    @property
+    def arithmetic_depth(self) -> int:
+        """Arithmetic depth of the operator."""
+        return 0
+
+    @property
+    def is_verified_hermitian(self) -> bool:
```
Comment (astralcai) [inline pennylane/operation2.py:280]: I wonder if we could get rid of this attribute now in favour of registering dispatches for `qp.is_hermitian`. @albi3ro

File: pennylane/operation2.py (hunk 5 above)
Comment (mudit2812) [reply pennylane/operation2.py:204]: These make structured access to operator inputs much easier, and will likely be used in places outside `Operator2`, especially for integration. Private and protected variables should strictly be for values that shouldn't be accessed outside `Operator2`. I think that is a forgotten distinction in much of pennylane—public ≠ user-facing

File: pennylane/operation2.py
Comment (mudit2812) [inline pennylane/operation2.py:159]: A question I'm starting to wonder now - do we need to do this in the constructor?

Comment (astralcai) [reply pennylane/operation2.py:159]: Where else would we have it?

Comment (mudit2812) [reply pennylane/operation2.py:159]: I mean, I'm not sure if we need it at all. Subclasses that have pauli reps will have to override `pauli_rep` anyway, and we can just update the default `pauli_rep` to return `None` instead of `_pauli_rep`

Comment (albi3ro) [reply pennylane/operation2.py:159]: Set the default on the class.

```
class Operator2:

    _pauli_rep : qp.pauli.PauliSentence | None = None
   
```

File: pennylane/operation2.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/operation2.py:280]: 🤷

**Review by albi3ro — APPROVED** [review]

**Review by albi3ro — APPROVED** [review]

File: pennylane/operation2.py
Comment (mudit2812) [reply pennylane/operation2.py:159]: That also seems odd, because it should be an instance attribute since it depends on wires. Essentially, the question is, do we need `_pauli_rep` when we have `pauli_rep`?

Comment (albi3ro) [reply pennylane/operation2.py:159]:
```
return getattr(self, "_pauli_rep", None)
```
?

Comment (astralcai) [reply pennylane/operation2.py:159]: I see your point, we can simply have
```
@property
def pauli_rep(self):
    return None
```
and each op with its own pauli rep would just override this. There really is no reason for the base class to hold a private attribute called `_pauli_rep`.

File: pennylane/operation2.py [hunk 7]
```diff
@@ -165,6 +164,436 @@ def __init__(self, *args, **kwargs):
...
+        True
+
+        """
+        return False
+
+    @property
+    def pauli_rep(self) -> "qp.pauli.PauliSentence | None":
+        """A :class:`~.PauliSentence` representation of the Operator, or ``None``
+        if it doesn't have one."""
+        return getattr(self, "_pauli_rep", None)
```
Comment (astralcai) [inline pennylane/operation2.py]: We could just `return None` here, and then each individual op with an actual pauli rep would override this? I don't like that we're relying on subclasses to add a private attribute of a certain name so that a public property would work.

**Review by astralcai — COMMENTED** [review]
~Have we intentionally left out some methods? For example, I don't see `generator`, `diagonalizing_gates`, etc. Are they going to be added in a separate PR or have we made a conscious decision to leave them out?~

EDIT: I see the follow-up PR now.

File: pennylane/operation2.py [hunk 8]
```diff
@@ -165,6 +164,436 @@ def __init__(self, *args, **kwargs):
...
+        for n, wires in self.wire_args.items():
+            # Flattening/unflattening allows mapping hybrid wire arguments
+            leaves, tree = flatten(wires, is_leaf=lambda w: isinstance(w, Wires))
+            mapped_leaves = [Wires([wire_map.get(w, w) for w in leaf]) for leaf in leaves]
+            new_wires = unflatten(mapped_leaves, tree)
+            new_op._bound_args.arguments[n] = new_wires
+
+        if (p_rep := getattr(self, "_pauli_rep", None)) is not None:
+            # pylint: disable=attribute-defined-outside-init
+            new_op._pauli_rep = p_rep.map_wires(wire_map)
```
Comment (astralcai) [inline pennylane/operation2.py:501]: This no longer applies.

Comment (mudit2812) [reply pennylane/operation2.py:501]: I added it so potential subclasses don't need to override `map_wires` if they do have a private `pauli_rep`. Although maybe I could just do this for `pauli_rep` instead of `_pauli_rep`. What do you think?

Adding back `_pauli_rep` is also reasonable I think

Comment (astralcai) [reply pennylane/operation2.py:501]: Yeah I do see the merit of adding back `_pauli_rep` for performance.

Comment (astralcai) [reply pennylane/operation2.py:501]: Although perhaps instead of requiring subclasses to overshadow a private attribute, we can define a setter for the `pauli_rep` property and make `_pauli_rep` entirely internal to the `Operator2` base class.

Comment (mudit2812) [reply pennylane/operation2.py:501]: A setter would go against operators being immutable

Comment (astralcai) [reply pennylane/operation2.py:501]: ~would a sub-class ever need to modify `_pauli_rep`?~

Ok I guess it's probably just easier to have `_pauli_rep` as a private attribute then.

**Review by astralcai — COMMENTED** [review]
Let's add `_pauli_rep` back then. Otherwise happy to approve.

**Review by astralcai — APPROVED** [review]
Nice!

File: pennylane/operation2.py [hunk 9]
```diff
@@ -166,6 +169,435 @@ def __init__(self, *args, **kwargs):
...
+        ...     def adjoint(self):
+        ...         return self
+        ...
+        >>> op = MyClass(0.5, wires=0).adjoint()
+        >>> op
+        MyClass(phi=0.5, wires=[0])
+        """
+        raise AdjointUndefinedError
+
+    def map_wires(self, wire_map: dict[Hashable, Hashable]) -> "Operator2":
```
Comment (dime10) [inline pennylane/operation2.py:480]: Is the assumption here that we never touch the structure of the wire args?

**Review by dime10 — COMMENTED** [review]
no comments from my side 👌

File: pennylane/operation2.py (hunk 9 above)
Comment (mudit2812) [reply pennylane/operation2.py:480]: Why would we want to?

Comment (dime10) [reply pennylane/operation2.py:480]: All good, I think I was overthinking it 👌
